bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthClientServiceImpl.java

```diff
@@ -0,0 +1,381 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.auth.oauth2client.internal;
+
+import static org.eclipse.smarthome.core.auth.client.oauth2.Keyword.CODE;
+import static org.eclipse.smarthome.core.auth.client.oauth2.Keyword.STATE;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.Map;
+import java.util.UUID;
+
+import javax.crypto.Cipher;
+import javax.net.ssl.TrustManager;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.util.UrlEncoded;
+import org.eclipse.smarthome.core.auth.client.oauth2.CipherConfigurator;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthClientService;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthError;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthException;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthProviderParams;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthRequestParams;
+import org.eclipse.smarthome.core.auth.client.oauth2.OAuthToken;
+import org.eclipse.smarthome.core.storage.StorageService;
+import org.osgi.framework.BundleContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of OAuthClientService.  
+ * 
+ * It requires the following services:
+ *
+ * org.eclipse.smarthome.core.storage.Storage (mandatory; for storing grant tokens, access tokens and refresh tokens)
+ * 
+ * javax.net.ssl.TrustManager[] (optional; this is used to provide SSL options during HTTPS negotiations with
+ * authorization providers. i.e. during SSLContext.init(...) )
+ *
+ * The OAuthTokens, request parameters are stored and persisted using the "Storage" service.  
+ * This allows the token to be automatically refreshed when needed.
+ *
+ * 
+ *
+ * @author m.bock - Initial contribution
+ * @author Gary Tse - Initial contribution
+ *
+ */
+@Component(name = "org.eclipse.smarthome.auth.oauth2.client.OAuthClientService", servicefactory = true, property = {
```

Comment:
`configurationPid="org.eclipse.smarthome.any.pid", configurationPolicy=REQUIRE` would enable the creation of multiple instances of `OAuthClientServiceImpl`.

If we do this, and if the component has a configuration policy of `required` as shown above, then we will get multiple instances of the component, with each instance potentially receiving different configuration data. DS will continue to maintain all the service references in our component of course.